### PR TITLE
Add TraceProcessor

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
Adds a TraceProcessor to process traces before they get to a writer. This sets
up #2038 which will add the trace building logic to the processor.

## Checklist
- [x] ~Added to the correct milestone.~
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [x] ~Library documentation is updated.~
- [x] ~[Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).~

### CODE REVIEW

File: ddtrace/tracer.py [hunk 1]
```diff
@@ -283,15 +283,21 @@ def configure(
...
             if hasattr(self, "writer") and self.writer.is_alive():
                 self.writer.stop()
 
-            self.writer = AgentWriter(
-                url,
-                sampler=self.sampler,
-                priority_sampler=self.priority_sampler,
-                dogstatsd=get_dogstatsd_client(self._dogstatsd_url),
-                report_metrics=config.health_metrics_enabled,
+            self.writer = TraceProcessor(
```
Comment (P403n1x87) [inline ddtrace/tracer.py:292]: Given that there are both a `writer` and a `processor` modules now, I think `self.writer` becomes a misleading name because it now holds a reference to a processor rather than an actual writer. So, if the writer is now becoming an implementation detail, would it make sense to rename this to `self.processor` instead? I would then suggest that the processor implements a `run` method rather than `write`.

Comment (Kyle-Verhoog) [reply ddtrace/tracer.py:292]: Totally agreed. I don't think that we can do away with `tracer.writer` as I believe we have recommended to people to do `tracer.writer.flush_queue()` in the past. We could expose `writer` as a property that does `self.processor.writer` to deal with this though and would actually maintain the functionality better.

I'll make that change thanks @P403n1x87!

**Review by P403n1x87 — APPROVED** [review]

File: ddtrace/internal/processor.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+from ddtrace.filters import TraceFilter
+from ddtrace.internal.writer import TraceWriter
+
+from .logger import get_logger
+
+
+log = get_logger(__name__)
+
+
+class TraceProcessor(object):
```
Comment (jd) [inline ddtrace/internal/processor.py:13]: Use `attr`?

**Review by jd — CHANGES_REQUESTED** [review]
Looks like there is too many changes in this PR

**Comment by Kyle-Verhoog** [comment]
@jd yeah if we're good with the approach then I'm planning on splitting out the `configure`/`tracer.__init__` + test helper changes 🙂

File: ddtrace/internal/processor.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+from .logger import get_logger
+
+
+log = get_logger(__name__)
+
+
+@attr.s
+class TraceProcessor(object):
+    writer = attr.ib(type=TraceWriter, init=True)
+    _filters = attr.ib(type=List[TraceFilter], init=True)
```
Comment (jd) [inline ddtrace/internal/processor.py:17]: `init=True` is the default so no need to set it

File: ddtrace/internal/processor.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            try:
+                trace = filtr.process_trace(trace)
+            except Exception:
+                log.error("error applying filter %r to traces", filtr, exc_info=True)
+            else:
+                if trace is None:
+                    log.debug("dropping trace due to filter %r", filter)
+                    return
+
+        self.writer.write(trace)
```
Comment (jd) [inline ddtrace/internal/processor.py:31]: I think the API should return the `trace` and not use a `writer` here.
That'd make the processor reusable. It's then up to the caller to write the trace.

**Review by jd — CHANGES_REQUESTED** [review]

File: ddtrace/internal/processor.py (hunk 4 above)
Comment (Kyle-Verhoog) [reply ddtrace/internal/processor.py:31]: Great idea!

**Review by jd — COMMENTED** [review]
The `TraceProcessor` class is good to me, but there are a lot of other changes that _seems_ unrelated, no?

**Comment by Kyle-Verhoog** [comment]
Most of the changes exist in #2134 and will go away with the merge, however I can undo them in this PR now that it's pulled out!

File: tests/tracer/test_processor.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+from ddtrace import Span
+from ddtrace.filters import TraceFilter
+from ddtrace.internal.processor import TraceProcessor
+
+
+def test_no_filters():
+    tp = TraceProcessor([])
+    trace = [Span(None, "span1"), Span(None, "span2")]
+    spans = tp.process(trace)
+    assert spans == trace
```
Comment (P403n1x87) [inline tests/tracer/test_processor.py:12]: Would it make sense to check that a `TraceProcessor` with no filters yields the same input trace, rather than potentially a copy of it with the same content?

~~~ python
assert spans is trace
~~~

File: ddtrace/tracer.py [hunk 6]
```diff
@@ -608,24 +611,16 @@ def write(self, spans):
         Send the trace to the writer to enqueue the spans list in the agent
         sending queue.
         """
-        if not spans:
+        if not spans or not self.enabled:
```
Comment (majorgreys) [inline ddtrace/tracer.py:614]: Do we have a test that covers this change?

Comment (jd) [inline ddtrace/tracer.py:614]: This looks unrelated to the purpose of this patch, no?

File: tests/tracer/test_processor.py (hunk 5 above)
Comment (jd) [reply tests/tracer/test_processor.py:12]: I don't think it's part of the assumption/protocol of this class, so I'd say the current test is correct.

**Review by jd — CHANGES_REQUESTED** [review]

File: ddtrace/tracer.py (hunk 6 above)
Comment (Kyle-Verhoog) [reply ddtrace/tracer.py:614]: Hm, it was already being done previously but it was done later. I'll move it back to where it was originally.

Comment (Kyle-Verhoog) [reply ddtrace/tracer.py:614]: Hmm it was meant to introduce any new functionality as it was already being done below. As per the other comment, I'll move it back to where it was originally.

File: tests/tracer/test_processor.py (hunk 5 above)
Comment (Kyle-Verhoog) [reply tests/tracer/test_processor.py:12]: Yeah I don't think we have to make this assumption

**Review by majorgreys — APPROVED** [review]

**Review by jd — APPROVED** [review]
